### ISSUE TITLE
faddeeva_erfc and friends not handled for macOS aarch64

### ISSUE THREAD

**Issue by motabbara** [EnzymeAD/Enzyme#1327]
Julia SpecialFunctions testing breaks on unhandled CFP assert with macOS aarch64 for complex inputs as tablegen assumes inputs are passed as structs. On macOS aarch64 these are arrays, similar to how cabs is currently handled for complex inputs. I couldn't figure out how to write the rule in tablegen.

**Comment by motabbara** [comment]
```
define void @julia_erf_140([2 x double]* noalias nocapture noundef nonnull sret([2 x double]) align 8 dereferenceable(16) %0, [2 x double]* nocapture noundef nonnull readonly align 8 dereferenceable(16) %1) #0 {
top:
  %gcframe7 = alloca [3 x {}*], align 16
  %gcframe7.sub = getelementptr inbounds [3 x {}*], [3 x {}*]* %gcframe7, i64 0, i64 0
  %2 = bitcast [3 x {}*]* %gcframe7 to i8*
  call void @llvm.memset.p0i8.i32(i8* noundef nonnull align 16 dereferenceable(24) %2, i8 0, i32 24, i1 false)
  %3 = call {}*** inttoptr (i64 6822325776 to {}*** (i64)*)(i64 262) #6
; ┌ @ /Users/mtabbara/.julia/packages/SpecialFunctions/sKqs4/src/erf.jl:19 within `_erf`
   %4 = bitcast [3 x {}*]* %gcframe7 to i64*
   store i64 4, i64* %4, align 16
   %5 = load {}**, {}*** %3, align 8
   %6 = getelementptr inbounds [3 x {}*], [3 x {}*]* %gcframe7, i64 0, i64 1
   %7 = bitcast {}** %6 to {}***
   store {}** %5, {}*** %7, align 8
   %8 = bitcast {}*** %3 to {}***
   store {}** %gcframe7.sub, {}*** %8, align 8
   %.elt = getelementptr inbounds [2 x double], [2 x double]* %1, i64 0, i64 0
   %.unpack = load double, double* %.elt, align 8
   %.elt4 = getelementptr inbounds [2 x double], [2 x double]* %1, i64 0, i64 1
   %.unpack5 = load double, double* %.elt4, align 8
   %9 = load atomic void ()*, void ()** @libname_Faddeeva_erf_142 unordered, align 8
   %.not = icmp eq void ()* %9, null
   br i1 %.not, label %dlsym, label %ccall

dlsym:                                            ; preds = %top
   %10 = load atomic {}*, {}** inttoptr (i64 4723807120 to {}**) unordered, align 16
   %11 = getelementptr inbounds [3 x {}*], [3 x {}*]* %gcframe7, i64 0, i64 2
   store {}* %10, {}** %11, align 16
   %12 = call void ()* @ijl_lazy_load_and_lookup({}* %10, i8* getelementptr inbounds ([13 x i8], [13 x i8]* @_j_str1, i64 0, i64 0))
   store atomic void ()* %12, void ()** @libname_Faddeeva_erf_142 release, align 8
   br label %ccall

ccall:                                            ; preds = %dlsym, %top
   %13 = phi void ()* [ %9, %top ], [ %12, %dlsym ]
   %14 = insertvalue [2 x double] zeroinitializer, double %.unpack, 0
   %15 = insertvalue [2 x double] %14, double %.unpack5, 1
   %16 = bitcast void ()* %13 to [2 x double] ([2 x double], double)*
   %17 = call [2 x double] %16([2 x double] %15, double 0.000000e+00)
   %.fca.0.extract = extractvalue [2 x double] %17, 0
   %.fca.1.extract = extractvalue [2 x double] %17, 1
; └
  %.sroa.0.0..sroa_idx = getelementptr inbounds [2 x double], [2 x double]* %0, i64 0, i64 0
  store double %.fca.0.extract, double* %.sroa.0.0..sroa_idx, align 8
  %.sroa.2.0..sroa_idx1 = getelementptr inbounds [2 x double], [2 x double]* %0, i64 0, i64 1
  store double %.fca.1.extract, double* %.sroa.2.0..sroa_idx1, align 8
  %18 = load {}*, {}** %6, align 8
  %19 = bitcast {}*** %3 to {}**
  store {}* %18, {}** %19, align 8
  ret void
}
```

**Comment by wsmoses** [comment]
Oh that's not a Tablegen issue, but that library call isn't being inlined by Enzyme for some reason?

**Comment by motabbara** [comment]
This is before Enzyme gets to it. I think the issue is that I have 

`%17 = call [2 x double] %16([2 x double] %15, double 0.000000e+00)`

on aarch64 versus 

`%17 = call { double, double }  %16({ double, double }  %15, double 0.000000e+00)`

on x86_64.

**Comment by wsmoses** [comment]
Would you be able to produce an llvm test that hits it by chance? E.g. I would do a show string(mod). at the start of enzyme!, then add the autodiff call to the function of interest.

**Comment by motabbara** [comment]
`
; ModuleID = 'start'source_filename = "start"
target datalayout = "e-m:o-i64:64-i128:128-n32:64-S128-ni:10:11:12:13"
target triple = "arm64-apple-darwin22.4.0"

@libname_Faddeeva_erf_2215 = local_unnamed_addr global void ()* null

; Function Attrs: nofree readnone
declare {}*** @julia.get_pgcstack() local_unnamed_addr #0

; Function Attrs: nounwind readnone
declare [2 x double] @Faddeeva_erf([2 x double], double) local_unnamed_addr #1

define [2 x double] @julia_erf_2213_inner.1([2 x double] %0) local_unnamed_addr #2 !dbg !4 {
entry:
  %1 = alloca [2 x double], align 8, !dbg !7
  %2 = addrspacecast [2 x double]* %1 to [2 x double] addrspace(11)*, !dbg !7
  %.fca.0.extract = extractvalue [2 x double] %0, 0, !dbg !7
  %.fca.0.gep = getelementptr inbounds [2 x double], [2 x double]* %1, i64 0, i64 0, !dbg !7
  store double %.fca.0.extract, double* %.fca.0.gep, align 8, !dbg !7, !noalias !8
  %.fca.1.extract = extractvalue [2 x double] %0, 1, !dbg !7
  %.fca.1.gep = getelementptr inbounds [2 x double], [2 x double]* %1, i64 0, i64 1, !dbg !7
  store double %.fca.1.extract, double* %.fca.1.gep, align 8, !dbg !7, !noalias !8
  %3 = call {}*** @julia.get_pgcstack(), !noalias !11
  %4 = call [2 x double] @Faddeeva_erf([2 x double] %0, double noundef 0.000000e+00) [ "jl_roots"([2 x double] addrspace(11)* %2) ], !dbg !14, !noalias !11
  ret [2 x double] %4, !dbg !7
}

attributes #0 = { nofree readnone "enzyme_inactive" "enzyme_shouldrecompute" "enzymejl_world"="33510" }
attributes #1 = { nounwind readnone }
attributes #2 = { "enzymejl_mi"="10968903968" "enzymejl_rt"="4817123296" "enzymejl_world"="33510" "probe-stack"="inline-asm" }

!llvm.module.flags = !{!0, !1}
!llvm.dbg.cu = !{!2}

!0 = !{i32 2, !"Dwarf Version", i32 4}
!1 = !{i32 2, !"Debug Info Version", i32 3}
!2 = distinct !DICompileUnit(language: DW_LANG_Julia, file: !3, producer: "julia", isOptimized: true, runtimeVersion: 0, emissionKind: LineTablesOnly, nameTableKind: None)
!3 = !DIFile(filename: "/Users/mtabbara/.julia/packages/SpecialFunctions/sKqs4/src/erf.jl", directory: ".")
!4 = distinct !DISubprogram(name: "erf", linkageName: "julia_erf_2213", scope: null, file: !3, line: 13, type: !5, scopeLine: 13, spFlags: DISPFlagDefinition | DISPFlagOptimized, unit: !2, retainedNodes: !6)
!5 = !DISubroutineType(types: !6)
!6 = !{}
!7 = !DILocation(line: 0, scope: !4)
!8 = !{!9}
!9 = distinct !{!9, !10, !"na_addr13"}
!10 = distinct !{!10, !"addr13"}
!11 = !{!12}
!12 = distinct !{!12, !13, !"julia_erf_2213_inner: argument 0"}
!13 = distinct !{!13, !"julia_erf_2213_inner"}
!14 = !DILocation(line: 19, scope: !15, inlinedAt: !16)
!15 = distinct !DISubprogram(name: "_erf;", linkageName: "_erf", scope: !3, file: !3, type: !5, spFlags: DISPFlagDefinition | DISPFlagOptimized, unit: !2, retainedNodes: !6)
!16 = distinct !DILocation(line: 13, scope: !4, inlinedAt: !17)
!17 = distinct !DILocation(line: 0, scope: !4)

`

**Comment by motabbara** [comment]
`
; ModuleID = 'start'
source_filename = "start"
target datalayout = "e-m:o-i64:64-i128:128-n32:64-S128-ni:10:11:12:13"
target triple = "arm64-apple-darwin22.4.0"

@libname_Faddeeva_erf_2215 = local_unnamed_addr global void ()* null

; Function Attrs: nofree readnone
declare {}*** @julia.get_pgcstack() local_unnamed_addr #0

; Function Attrs: nounwind readnone
declare [2 x double] @Faddeeva_erf([2 x double], double) local_unnamed_addr #1

define [2 x double] @julia_erf_2213_inner.1([2 x double] %0) local_unnamed_addr #2 !dbg !4 {
entry:
  %1 = alloca [2 x double], align 8, !dbg !7
  %2 = addrspacecast [2 x double]* %1 to [2 x double] addrspace(11)*, !dbg !7
  %.fca.0.extract = extractvalue [2 x double] %0, 0, !dbg !7
  %.fca.0.gep = getelementptr inbounds [2 x double], [2 x double]* %1, i64 0, i64 0, !dbg !7
  store double %.fca.0.extract, double* %.fca.0.gep, align 8, !dbg !7, !noalias !8
  %.fca.1.extract = extractvalue [2 x double] %0, 1, !dbg !7
  %.fca.1.gep = getelementptr inbounds [2 x double], [2 x double]* %1, i64 0, i64 1, !dbg !7
  store double %.fca.1.extract, double* %.fca.1.gep, align 8, !dbg !7, !noalias !8
  %3 = call {}*** @julia.get_pgcstack(), !noalias !11
  %4 = call [2 x double] @Faddeeva_erf([2 x double] %0, double noundef 0.000000e+00) [ "jl_roots"([2 x double] addrspace(11)* %2) ], !dbg !14, !noalias !11
  ret [2 x double] %4, !dbg !7
}

attributes #0 = { nofree readnone "enzyme_inactive" "enzyme_shouldrecompute" "enzymejl_world"="33510" }
attributes #1 = { nounwind readnone }
attributes #2 = { "enzymejl_mi"="10968903968" "enzymejl_rt"="4817123296" "enzymejl_world"="33510" "probe-stack"="inline-asm" }

!llvm.module.flags = !{!0, !1}
!llvm.dbg.cu = !{!2}

!0 = !{i32 2, !"Dwarf Version", i32 4}
!1 = !{i32 2, !"Debug Info Version", i32 3}
!2 = distinct !DICompileUnit(language: DW_LANG_Julia, file: !3, producer: "julia", isOptimized: true, runtimeVersion: 0, emissionKind: LineTablesOnly, nameTableKind: None)
!3 = !DIFile(filename: "/Users/mtabbara/.julia/packages/SpecialFunctions/sKqs4/src/erf.jl", directory: ".")
!4 = distinct !DISubprogram(name: "erf", linkageName: "julia_erf_2213", scope: null, file: !3, line: 13, type: !5, scopeLine: 13, spFlags: DISPFlagDefinition | DISPFlagOptimized, unit: !2, retainedNodes: !6)
!5 = !DISubroutineType(types: !6)
!6 = !{}
!7 = !DILocation(line: 0, scope: !4)
!8 = !{!9}
!9 = distinct !{!9, !10, !"na_addr13"}
!10 = distinct !{!10, !"addr13"}
!11 = !{!12}
!12 = distinct !{!12, !13, !"julia_erf_2213_inner: argument 0"}
!13 = distinct !{!13, !"julia_erf_2213_inner"}
!14 = !DILocation(line: 19, scope: !15, inlinedAt: !16)
!15 = distinct !DISubprogram(name: "_erf;", linkageName: "_erf", scope: !3, file: !3, type: !5, spFlags: DISPFlagDefinition | DISPFlagOptimized, unit: !2, retainedNodes: !6)
!16 = distinct !DILocation(line: 13, scope: !4, inlinedAt: !17)
!17 = distinct !DILocation(line: 0, scope: !4)
`

**Comment by motabbara** [comment]
; ModuleID = 'start'
source_filename = "start"
target datalayout = "e-m:o-i64:64-i128:128-n32:64-S128-ni:10:11:12:13"
target triple = "arm64-apple-darwin22.4.0"

@libname_Faddeeva_erf_2215 = local_unnamed_addr global void ()* null

; Function Attrs: nofree readnone
declare {}*** @julia.get_pgcstack() local_unnamed_addr #0

; Function Attrs: nounwind readnone
declare [2 x double] @Faddeeva_erf([2 x double], double) local_unnamed_addr #1

define [2 x double] @julia_erf_2213_inner.1([2 x double] %0) local_unnamed_addr #2 !dbg !4 {
entry:
  %1 = alloca [2 x double], align 8, !dbg !7
  %2 = addrspacecast [2 x double]* %1 to [2 x double] addrspace(11)*, !dbg !7
  %.fca.0.extract = extractvalue [2 x double] %0, 0, !dbg !7
  %.fca.0.gep = getelementptr inbounds [2 x double], [2 x double]* %1, i64 0, i64 0, !dbg !7
  store double %.fca.0.extract, double* %.fca.0.gep, align 8, !dbg !7, !noalias !8
  %.fca.1.extract = extractvalue [2 x double] %0, 1, !dbg !7
  %.fca.1.gep = getelementptr inbounds [2 x double], [2 x double]* %1, i64 0, i64 1, !dbg !7
  store double %.fca.1.extract, double* %.fca.1.gep, align 8, !dbg !7, !noalias !8
  %3 = call {}*** @julia.get_pgcstack(), !noalias !11
  %4 = call [2 x double] @Faddeeva_erf([2 x double] %0, double noundef 0.000000e+00) [ "jl_roots"([2 x double] addrspace(11)* %2) ], !dbg !14, !noalias !11
  ret [2 x double] %4, !dbg !7
}

attributes #0 = { nofree readnone "enzyme_inactive" "enzyme_shouldrecompute" "enzymejl_world"="33510" }
attributes #1 = { nounwind readnone }
attributes #2 = { "enzymejl_mi"="10968903968" "enzymejl_rt"="4817123296" "enzymejl_world"="33510" "probe-stack"="inline-asm" }

!llvm.module.flags = !{!0, !1}
!llvm.dbg.cu = !{!2}

!0 = !{i32 2, !"Dwarf Version", i32 4}
!1 = !{i32 2, !"Debug Info Version", i32 3}
!2 = distinct !DICompileUnit(language: DW_LANG_Julia, file: !3, producer: "julia", isOptimized: true, runtimeVersion: 0, emissionKind: LineTablesOnly, nameTableKind: None)
!3 = !DIFile(filename: "/Users/mtabbara/.julia/packages/SpecialFunctions/sKqs4/src/erf.jl", directory: ".")
!4 = distinct !DISubprogram(name: "erf", linkageName: "julia_erf_2213", scope: null, file: !3, line: 13, type: !5, scopeLine: 13, spFlags: DISPFlagDefinition | DISPFlagOptimized, unit: !2, retainedNodes: !6)
!5 = !DISubroutineType(types: !6)
!6 = !{}
!7 = !DILocation(line: 0, scope: !4)
!8 = !{!9}
!9 = distinct !{!9, !10, !"na_addr13"}
!10 = distinct !{!10, !"addr13"}
!11 = !{!12}
!12 = distinct !{!12, !13, !"julia_erf_2213_inner: argument 0"}
!13 = distinct !{!13, !"julia_erf_2213_inner"}
!14 = !DILocation(line: 19, scope: !15, inlinedAt: !16)
!15 = distinct !DISubprogram(name: "_erf;", linkageName: "_erf", scope: !3, file: !3, type: !5, spFlags: DISPFlagDefinition | DISPFlagOptimized, unit: !2, retainedNodes: !6)
!16 = distinct !DILocation(line: 13, scope: !4, inlinedAt: !17)
!17 = distinct !DILocation(line: 0, scope: !4)

**Comment by motabbara** [comment]
Fixed by #1331
